### PR TITLE
Add keyword splat type to keyword arguments descriptor

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

File: src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java [hunk 1]
```diff
@@ -16,14 +16,27 @@
  * arguments were passed. The callee must handle that for now. */
 public class KeywordArgumentsDescriptor extends ArgumentsDescriptor {
 
+    public enum SplatType {
+        NO_SPLAT,
+        ONLY_SPLAT,
+        PRE_SPLAT,
+        POST_SPLAT
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:23]: Is it useful to have that amount of details? Not against it, but wondering.

What's used for:
```ruby
kw = { b: 2 }
p(a: 1, **kw, c: 3)
# => {:a=>1, :b=>2, :c=>3}
```

File: spec/truffle/arguments_descriptor_spec.rb [hunk 2]
```diff
@@ -102,45 +102,73 @@ def posn_kws_defaults(a, b: 101, c: 102)
...
+    def kwargs(**kwargs)
+      info(kwargs, Primitive.arguments_descriptor, Primitive.arguments)
+    end
+
+    splatted = {b: 3}
+    info = kwargs(**splatted)
+    info.values.should_not.equal?(splatted)
+    info.descriptor.should == [:only_splat] if truffleruby?
+    info.arguments[0].should.equal?(splatted) if truffleruby?
+  end
```
Comment (eregon) [inline spec/truffle/arguments_descriptor_spec.rb:162]: Could you add a spec for a splat in the middle? `kwargs(a: 1, **kw, c: 3)`

Comment (eregon) [inline spec/truffle/arguments_descriptor_spec.rb:162]: Could you add a spec(s) for a splat in the middle? `kwargs(a: 1, **kw, c: 3)`

File: src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java [hunk 3]
```diff
@@ -16,14 +16,27 @@
  * arguments were passed. The callee must handle that for now. */
 public class KeywordArgumentsDescriptor extends ArgumentsDescriptor {
 
+    public enum SplatType {
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:19]: Maybe should be KeywordSplatType, if imported statically anywhere.
`SplatType` on its own sounds like a `*ary` splat.

Comment (eregon) [inline src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:19]: Maybe should be KeywordSplatType, if imported statically anywhere.
`SplatType` on its own sounds like a `*ary` splat.

File: src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java (hunk 1 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:23]: Pre and post are needed (as the newly added specs show) to know when to override explicit keyword arguments and when not to.

File: spec/truffle/arguments_descriptor_spec.rb (hunk 2 above)
Comment (eregon) [reply spec/truffle/arguments_descriptor_spec.rb:199]: I think one way to handle this is to either consider all of it is a splat ONLY_SPLAT and in the translator already introduce the nodes to merge and have a single rest kw arg, or e.g. consider it a PRE_SPLAT and in the translator merge the trailer in the kw splat.

File: src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:23]: One way to simplify (to not have to deal with this complexity on the callee side) is for all kw splats then to let the caller merge into a single RubyHash, and have no literal keywords in the descriptor used for them.
I'm not sure how frequent is a mix of literal and rest kwargs.

Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/arguments/KeywordArgumentsDescriptor.java:23]: I think pre and post are easy so I'll try to accommodate them.

**Comment by chrisseaton** [comment]
We'll come back to this later with a new PR.

**Pull request closed without merging**